Code generator produces C# code which fails to compile

Protocol file:
https://github.com/google/sentencepiece/blob/master/src/sentencepiece_model.proto

Compiler output:
sentencepiece_model.generated.cs(183,20,183,50): error CS0266: Cannot implicitly convert type 'double' to 'float'. An explicit conversion exists (are you missing a cast?)

The generated code:
```
[global::ProtoBuf.ProtoMember(10, Name = @"character_coverage")]
[global::System.ComponentModel.DefaultValue(0.9995)]
public float CharacterCoverage
{
    get => __pbn__CharacterCoverage ?? 0.9995;
    set => __pbn__CharacterCoverage = value;
}
```

I'm using .NET 6.

I’ve tried to workaround by replacing the number in the `.proto` from `0.9995` to ` 0.9995f`, but this time it refused to compile the protocol, saying `Error PBN1008: invalid floating-point number`

--- a/docs/releasenotes.md
+++ b/docs/releasenotes.md
@@ -14,6 +14,7 @@ Packages are available on NuGet: [protobuf-net](https://www.nuget.org/packages/p
 
 ## unreleased
 
+- fix bug with default values not including literal suffixes (#1037)
 - fix SDKs used by BuildTools[.Legacy]
 
 ## 3.2.8

--- a/src/Examples/Issues/SO7333233.cs
+++ b/src/Examples/Issues/SO7333233.cs
@@ -102,9 +102,10 @@ namespace Examples.Issues
         [Fact]
         public  void Execute()
         {
+            var model = RuntimeTypeModel.Create();
             // note these are unrelated networks, so we can use the same field-numbers
-            RuntimeTypeModel.Default[typeof(IRule<Ant>)].AddSubType(1, typeof(AntRule1)).AddSubType(2, typeof(AntRule2));
-            RuntimeTypeModel.Default[typeof(IRule<Cat>)].AddSubType(1, typeof(CatRule1)).AddSubType(2, typeof(CatRule2));
+            model[typeof(IRule<Ant>)].AddSubType(1, typeof(AntRule1)).AddSubType(2, typeof(AntRule2));
+            model[typeof(IRule<Cat>)].AddSubType(1, typeof(CatRule1)).AddSubType(2, typeof(CatRule2));
 
             var antRules = new List<IRule<Ant>>();
             antRules.Add(new AntRule1());
@@ -116,7 +117,7 @@ namespace Examples.Issues
 
             using (var fs = File.Create(@"antRules.bin"))
             {
-                ProtoBuf.Serializer.Serialize(fs, antRules);
+                model.Serialize(fs, antRules);
 
                 fs.Close();
             }
@@ -124,14 +125,14 @@ namespace Examples.Issues
             using (var fs = File.OpenRead(@"antRules.bin"))
             {
                 List<IRule<Ant>> list;
-                list = ProtoBuf.Serializer.Deserialize<List<IRule<Ant>>>(fs);
+                list = model.Deserialize<List<IRule<Ant>>>(fs);
 
                 fs.Close();
             }
 
             using (var fs = File.Create(@"catRules.bin"))
             {
-                ProtoBuf.Serializer.Serialize(fs, catRules);
+                model.Serialize(fs, catRules);
 
                 fs.Close();
             }
@@ -139,7 +140,7 @@ namespace Examples.Issues
             using (var fs = File.OpenRead(@"catRules.bin"))
             {
                 List<IRule<Cat>> list;
-                list = ProtoBuf.Serializer.Deserialize<List<IRule<Cat>>>(fs);
+                list = model.Deserialize<List<IRule<Cat>>>(fs);
 
                 fs.Close();
             }

--- a/src/protobuf-net.Reflection.Test/Schemas/SchemaTests.cs
+++ b/src/protobuf-net.Reflection.Test/Schemas/SchemaTests.cs
@@ -289,6 +289,7 @@ namespace ProtoBuf.Schemas
             {
                 var defaultValueAttribute = (DefaultValueAttribute)Attribute.GetCustomAttribute(property, typeof(DefaultValueAttribute));
                 Assert.NotNull(defaultValueAttribute);
+                Assert.Equal(typeof(ulong), defaultValueAttribute.Value.GetType());
                 Assert.Equal(18446744073709551615UL, defaultValueAttribute.Value);
             }
         }

--- a/src/protobuf-net.Reflection/CSharpCodeGenerator.cs
+++ b/src/protobuf-net.Reflection/CSharpCodeGenerator.cs
@@ -391,7 +391,12 @@ namespace ProtoBuf.Reflection
                             case "inf": defaultValue = "double.PositiveInfinity"; break;
                             case "-inf": defaultValue = "double.NegativeInfinity"; break;
                             case "nan": defaultValue = "double.NaN"; break;
-                            default: defaultValue += "d"; break;
+                            default:
+                                if (!string.IsNullOrWhiteSpace(defaultValue))
+                                {
+                                    defaultValue += "d";
+                                }
+                                break;
                         }
                         break;
                     case FieldDescriptorProto.Type.TypeFloat:
@@ -400,21 +405,35 @@ namespace ProtoBuf.Reflection
                             case "inf": defaultValue = "float.PositiveInfinity"; break;
                             case "-inf": defaultValue = "float.NegativeInfinity"; break;
                             case "nan": defaultValue = "float.NaN"; break;
-                            default: defaultValue += "f"; break;
+                            default:
+                                if (!string.IsNullOrWhiteSpace(defaultValue))
+                                {
+                                    defaultValue += "f";
+                                }
+                                break;
                         }
                         break;
                     case FieldDescriptorProto.Type.TypeSfixed64:
                     case FieldDescriptorProto.Type.TypeSint64:
                     case FieldDescriptorProto.Type.TypeInt64:
-                        defaultValue += "l";
+                        if (!string.IsNullOrWhiteSpace(defaultValue))
+                        {
+                            defaultValue += "l";
+                        }
                         break;
                     case FieldDescriptorProto.Type.TypeFixed64:
                     case FieldDescriptorProto.Type.TypeUint64:
-                        defaultValue += "ul";
+                        if (!string.IsNullOrWhiteSpace(defaultValue))
+                        {
+                            defaultValue += "ul";
+                        }
                         break;
                     case FieldDescriptorProto.Type.TypeFixed32:
                     case FieldDescriptorProto.Type.TypeUint32:
-                        defaultValue += "u";
+                        if (!string.IsNullOrWhiteSpace(defaultValue))
+                        {
+                            defaultValue += "u";
+                        }
                         break;
                     case FieldDescriptorProto.Type.TypeEnum:
                         var enumType = ctx.TryFind<EnumDescriptorProto>(obj.TypeName);

--- a/src/protobuf-net.Reflection/CSharpCodeGenerator.cs
+++ b/src/protobuf-net.Reflection/CSharpCodeGenerator.cs
@@ -379,59 +379,73 @@ namespace ProtoBuf.Reflection
             {
                 defaultValue = obj.DefaultValue;
 
-                if (obj.type == FieldDescriptorProto.Type.TypeString)
+                switch (obj.type)
                 {
-                    defaultValue = string.IsNullOrEmpty(defaultValue) ? "\"\""
+                    case FieldDescriptorProto.Type.TypeString:
+                        defaultValue = string.IsNullOrEmpty(defaultValue) ? "\"\""
                         : ("@\"" + (defaultValue ?? "").Replace("\"", "\"\"") + "\"");
-                }
-                else if (obj.type == FieldDescriptorProto.Type.TypeDouble)
-                {
-                    switch (defaultValue)
-                    {
-                        case "inf": defaultValue = "double.PositiveInfinity"; break;
-                        case "-inf": defaultValue = "double.NegativeInfinity"; break;
-                        case "nan": defaultValue = "double.NaN"; break;
-                    }
-                }
-                else if (obj.type == FieldDescriptorProto.Type.TypeFloat)
-                {
-                    switch (defaultValue)
-                    {
-                        case "inf": defaultValue = "float.PositiveInfinity"; break;
-                        case "-inf": defaultValue = "float.NegativeInfinity"; break;
-                        case "nan": defaultValue = "float.NaN"; break;
-                    }
-                }
-                else if (obj.type == FieldDescriptorProto.Type.TypeEnum)
-                {
-                    var enumType = ctx.TryFind<EnumDescriptorProto>(obj.TypeName);
-                    if (enumType != null)
-                    {
-                        EnumValueDescriptorProto found = null;
-                        if (!string.IsNullOrEmpty(defaultValue))
+                        break;
+                    case FieldDescriptorProto.Type.TypeDouble:
+                        switch (defaultValue)
                         {
-                            found = enumType.Values.Find(x => x.Name == defaultValue);
+                            case "inf": defaultValue = "double.PositiveInfinity"; break;
+                            case "-inf": defaultValue = "double.NegativeInfinity"; break;
+                            case "nan": defaultValue = "double.NaN"; break;
+                            default: defaultValue += "d"; break;
                         }
-                        else if (ctx.Syntax == FileDescriptorProto.SyntaxProto2)
+                        break;
+                    case FieldDescriptorProto.Type.TypeFloat:
+                        switch (defaultValue)
                         {
-                            // find the first one; if that is a zero, we don't need it after all
-                            found = enumType.Values.FirstOrDefault();
-                            if (found != null && found.Number == 0)
+                            case "inf": defaultValue = "float.PositiveInfinity"; break;
+                            case "-inf": defaultValue = "float.NegativeInfinity"; break;
+                            case "nan": defaultValue = "float.NaN"; break;
+                            default: defaultValue += "f"; break;
+                        }
+                        break;
+                    case FieldDescriptorProto.Type.TypeSfixed64:
+                    case FieldDescriptorProto.Type.TypeSint64:
+                    case FieldDescriptorProto.Type.TypeInt64:
+                        defaultValue += "l";
+                        break;
+                    case FieldDescriptorProto.Type.TypeFixed64:
+                    case FieldDescriptorProto.Type.TypeUint64:
+                        defaultValue += "ul";
+                        break;
+                    case FieldDescriptorProto.Type.TypeFixed32:
+                    case FieldDescriptorProto.Type.TypeUint32:
+                        defaultValue += "u";
+                        break;
+                    case FieldDescriptorProto.Type.TypeEnum:
+                        var enumType = ctx.TryFind<EnumDescriptorProto>(obj.TypeName);
+                        if (enumType != null)
+                        {
+                            EnumValueDescriptorProto found = null;
+                            if (!string.IsNullOrEmpty(defaultValue))
                             {
-                                if (!isOptional) found = null; // we don't need it after all
+                                found = enumType.Values.Find(x => x.Name == defaultValue);
+                            }
+                            else if (ctx.Syntax == FileDescriptorProto.SyntaxProto2)
+                            {
+                                // find the first one; if that is a zero, we don't need it after all
+                                found = enumType.Values.FirstOrDefault();
+                                if (found != null && found.Number == 0)
+                                {
+                                    if (!isOptional) found = null; // we don't need it after all
+                                }
+                            }
+                            // for proto3 the default is 0, so no need to do anything - GetValueOrDefault() will do it all
+
+                            if (found != null)
+                            {
+                                defaultValue = ctx.NameNormalizer.GetName(found);
+                            }
+                            if (!string.IsNullOrWhiteSpace(defaultValue))
+                            {
+                                defaultValue = typeName + "." + defaultValue;
                             }
                         }
-                        // for proto3 the default is 0, so no need to do anything - GetValueOrDefault() will do it all
-
-                        if (found != null)
-                        {
-                            defaultValue = ctx.NameNormalizer.GetName(found);
-                        }
-                        if (!string.IsNullOrWhiteSpace(defaultValue))
-                        {
-                            defaultValue = typeName + "." + defaultValue;
-                        }
-                    }
+                        break;
                 }
             }
 

--- a/src/protobuf-net.Reflection/CSharpCodeGenerator.cs
+++ b/src/protobuf-net.Reflection/CSharpCodeGenerator.cs
@@ -343,7 +343,7 @@ namespace ProtoBuf.Reflection
             var typeName = GetTypeName(ctx, field, out var dataFormat, out _, out _, out var isMap);
             bool trackPresence = TrackFieldPresence(ctx, field, oneOfs, out _);
 
-            string defaultValue = GetDefaultValue(ctx, field, typeName);
+            string defaultValue = GetDefaultValue(ctx, field, typeName, out var suffix);
 
             if (isRepeated)
             {
@@ -365,16 +365,16 @@ namespace ProtoBuf.Reflection
             {
                 if (!string.IsNullOrWhiteSpace(defaultValue))
                 {
-                    ctx.WriteLine($"{Escape(name)} = {defaultValue};");
+                    ctx.WriteLine($"{Escape(name)} = {defaultValue}{suffix};");
                 }
             }
         }
 
-        private string GetDefaultValue(GeneratorContext ctx, FieldDescriptorProto obj, string typeName)
+        private string GetDefaultValue(GeneratorContext ctx, FieldDescriptorProto obj, string typeName, out string suffix)
         {
             string defaultValue = null;
             bool isOptional = obj.label == FieldDescriptorProto.Label.LabelOptional;
-
+            suffix = "";
             if (isOptional || ctx.EmitRequiredDefaults || obj.type == FieldDescriptorProto.Type.TypeEnum)
             {
                 defaultValue = obj.DefaultValue;
@@ -391,12 +391,7 @@ namespace ProtoBuf.Reflection
                             case "inf": defaultValue = "double.PositiveInfinity"; break;
                             case "-inf": defaultValue = "double.NegativeInfinity"; break;
                             case "nan": defaultValue = "double.NaN"; break;
-                            default:
-                                if (!string.IsNullOrWhiteSpace(defaultValue))
-                                {
-                                    defaultValue += "d";
-                                }
-                                break;
+                            default: suffix = "d"; break;
                         }
                         break;
                     case FieldDescriptorProto.Type.TypeFloat:
@@ -405,35 +400,21 @@ namespace ProtoBuf.Reflection
                             case "inf": defaultValue = "float.PositiveInfinity"; break;
                             case "-inf": defaultValue = "float.NegativeInfinity"; break;
                             case "nan": defaultValue = "float.NaN"; break;
-                            default:
-                                if (!string.IsNullOrWhiteSpace(defaultValue))
-                                {
-                                    defaultValue += "f";
-                                }
-                                break;
+                            default: suffix = "f"; break;
                         }
                         break;
                     case FieldDescriptorProto.Type.TypeSfixed64:
                     case FieldDescriptorProto.Type.TypeSint64:
                     case FieldDescriptorProto.Type.TypeInt64:
-                        if (!string.IsNullOrWhiteSpace(defaultValue))
-                        {
-                            defaultValue += "l";
-                        }
+                        suffix = "l";
                         break;
                     case FieldDescriptorProto.Type.TypeFixed64:
                     case FieldDescriptorProto.Type.TypeUint64:
-                        if (!string.IsNullOrWhiteSpace(defaultValue))
-                        {
-                            defaultValue += "ul";
-                        }
+                        suffix = "ul";
                         break;
                     case FieldDescriptorProto.Type.TypeFixed32:
                     case FieldDescriptorProto.Type.TypeUint32:
-                        if (!string.IsNullOrWhiteSpace(defaultValue))
-                        {
-                            defaultValue += "u";
-                        }
+                        suffix = "u";
                         break;
                     case FieldDescriptorProto.Type.TypeEnum:
                         var enumType = ctx.TryFind<EnumDescriptorProto>(obj.TypeName);
@@ -498,7 +479,7 @@ namespace ProtoBuf.Reflection
 
             bool suppressDefaultAttribute = !isOptional;
             var typeName = GetTypeName(ctx, field, out var dataFormat, out var nullabilityType, out var compatibilityLevel, out var isMap);
-            string defaultValue = GetDefaultValue(ctx, field, typeName);
+            string defaultValue = GetDefaultValue(ctx, field, typeName, out var suffix);
 
             WriteDataFormatAttribute();
             WriteNullabilityTypeAttribute();
@@ -545,9 +526,8 @@ namespace ProtoBuf.Reflection
                     case FieldDescriptorProto.Type.TypeUint64:
                         ctx.WriteLine($"[global::System.ComponentModel.DefaultValue(typeof(ulong), \"{defaultValue}\")]");
                         break;
-
                     default:
-                        ctx.WriteLine($"[global::System.ComponentModel.DefaultValue({defaultValue})]");
+                        ctx.WriteLine($"[global::System.ComponentModel.DefaultValue({defaultValue}{suffix})]");
                         break;
                 }
             }
@@ -600,7 +580,7 @@ namespace ProtoBuf.Reflection
             }
             else if (oneOf is object)
             {
-                var defValue = string.IsNullOrWhiteSpace(defaultValue) ? (ctx.Supports(CSharp7_1) ? "default" : $"default({typeName})") : defaultValue;
+                var defValue = string.IsNullOrWhiteSpace(defaultValue) ? (ctx.Supports(CSharp7_1) ? "default" : $"default({typeName})") : (defaultValue + suffix);
                 var fieldName = GetOneOfFieldName(oneOf.OneOf);
                 var storage = oneOf.GetStorage(field.type, field.TypeName);
                 ctx.WriteLine($"{GetAccess(GetAccess(field))} {typeName} {Escape(name)}").WriteLine("{").Indent();
@@ -665,6 +645,7 @@ namespace ProtoBuf.Reflection
                 {
                     tw.Write(" ?? ");
                     tw.Write(defaultValue);
+                    tw.Write(suffix);
                 }
                 else if (!isRef)
                 {
@@ -692,7 +673,7 @@ namespace ProtoBuf.Reflection
             else
             {
                 tw = ctx.Write($"{GetAccess(GetAccess(field))} {typeName}{(IsNullableType(ctx, field, defaultValue, isOptional) ? "?" : "")} {Escape(name)} {{ get; set; }}");
-                if (!string.IsNullOrWhiteSpace(defaultValue) && ctx.Supports(CSharp6)) tw.Write($" = {defaultValue};");
+                if (!string.IsNullOrWhiteSpace(defaultValue) && ctx.Supports(CSharp6)) tw.Write($" = {defaultValue}{suffix};");
                 tw.WriteLine();
             }
             ctx.WriteLine();
@@ -721,7 +702,7 @@ namespace ProtoBuf.Reflection
             ctx.WriteLine($"{GetAccess(GetAccess(file))} static partial class {Escape(name)}").WriteLine("{").Indent();
         }
         /// <summary>
-        /// Ends an extgensions block
+        /// Ends an extensions block
         /// </summary>
         protected override void WriteExtensionsFooter(GeneratorContext ctx, FileDescriptorProto file, ref object state)
         {

--- a/src/protobuf-net.Reflection/Descriptor.cs
+++ b/src/protobuf-net.Reflection/Descriptor.cs
@@ -5,7 +5,7 @@
 // </auto-generated>
 
 #region Designer generated code
-#pragma warning disable CS0612, CS0618, CS1591, CS3021, IDE1006, RCS1036, RCS1057, RCS1085, RCS1192
+#pragma warning disable CS0612, CS0618, CS1591, CS3021, IDE0079, IDE1006, RCS1036, RCS1057, RCS1085, RCS1192
 namespace Google.Protobuf.Reflection
 {
 
@@ -1322,5 +1322,5 @@ namespace Google.Protobuf.Reflection
 
 }
 
-#pragma warning restore CS0612, CS0618, CS1591, CS3021, IDE1006, RCS1036, RCS1057, RCS1085, RCS1192
+#pragma warning restore CS0612, CS0618, CS1591, CS3021, IDE0079, IDE1006, RCS1036, RCS1057, RCS1085, RCS1192
 #endregion

